doc/doxygen/public/developer_cpp_guide.doxygen

```diff
@@ -0,0 +1,163 @@
+
+
+
+@section cpp_guide_logging Logging
+To make direct output to `std::out` and `std::err` more consistent, %OpenMS provides several low-level macros:
+@code
+OPENMS_LOG_FATAL_ERROR, 
+OPENMS_LOG_ERROR
+OPENMS_LOG_WARN,
+OPENMS_LOG_INFO and
+OPENMS_LOG_DEBUG
+@endcode
+which should be used instead of the less descriptive `std::out` and `std::err` streams. 
+Furthermore, the OpenMS loggers insert console coloring for their output and have a deduplication cache build in, which prevents repetitive outputs by aggregating and counting their occurence.
+
+If you are writing an %OpenMS tool, you can also use the ProgressLogger to indicate how many percent of the processing has already been performed:
+<br><br>
+<b>Example: Tutorial_Logger.cpp</b>
+<br>
+Logging the Tool Progress
+
+\snippet Tutorial_Logger.cpp Logger
+
+Depending on how the user configures the ProgressLogger, this output is written to the command line or a log file.
+
+
+@section cpp_guide_input_output Input/Output
+
+Code like <code>stream_object << "example" << std::endl;</code> forces the output buffer to be flushed, i.e. written to disk/console immediately, which can be a big performance loss. Get used to writing code like <code>stream_object << "example\n";</code>. Debugging output can be an exception, because the content of the stream buffer may be lost upon segfault etc..
+
+Write many digits to avoid unnecessary rounding errors. In particular, using standard output stream operators, i.e. << for doubles and floats should be avoided when full precision is required because by default, not all significant digits will be written. Before you start using <code>os.precision(writtenDigits(FloatingPointType()));</code> and alike, it is strongly advised to convert to an <code>OpenMS::String</code>, i.e. <code>os << String(my_number)</code> because it's faster, and gives you all significant digits for each type (6 digits for <code>float</code>, 15 for <code>double</code>). Similarly, input stream operators are also slow, especially in VisualStudio, so switching to <code>OpenMS::String::toDouble()</code> is advised for performance reasons. If you do not need all significant digits, simply invoke <code>String(my_number, full_precision = false)</code> to get up to only three fractional digits for <code>float</code> and <code>double</code> types. For <code>Integer</code> types, there is no problem with streams, but again: <code>OpenMS::String(int i)</code> is faster. There is usually no heap allocation overhead for strings because of Small String Optimizations (SSO).
+
+
+
+@section cpp_guide_uint `UInt` vs. `Size`
+
+OpenMS uses some custom type definitions for simple arithmetic types, such as `UInt` (shorthand for `unsigned int`).
+When working with <code>STL</code> types (especially vectors), assign the return value of a <code>.size()</code> operation to the OpenMS type <code>Size</code>, which is defined as follows:
+
+@code{.cpp}
+// OpenMS/include/CONCEPT/Types.h
+typedef size_t Size;
```

Comment:
yes true, I was thinking if replacing our UInt32 etc is worthwhile, since its effectively an alias to uint32_t these days.
But that is for another day.

Reply:
other c++ projects, e.g. serenityOS, go with `u32` as shorthand. I can see benefits and potential issues there...
Maybe worth discussing during dev meeting.